[flat_tensor] Persist FreeableBuffers of external constants in method

lucylq opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #8437

## Problem
Currently, the FlatTensorDataMap persists tensors, and returns a FreeableBuffer with an empty free function.

The NamedDataMap should not persist data, as most cases (eg. delegate) will want it to be freed.

Ownership should be on the caller; `get_data` returns a FreeableBuffer that 'owns' the data. The FreeableBuffer in turn is owned by the caller.

NOTE: this doesn't support the case where we want to share plain tensors between methods/pte files at runtime. A custom NDM could support that use-case.

## This diff:
1. Introduces a 'NamedData' struct to method.h. This holds a key and a FreeeableBuffer.
2. Iterate over all the flatbuffer tensors to count the constants tagged with EXTERNAL.
NOTE: this will increase load time for all users. Potentially allocate chunks of 16 and use a linked list to store external constants, or store this number in PTE file.
3. Allocate space for num_external_constants using the method allocator.
4. Iterate over all flatbuffer tensors and use the named_data_map to resolve EXTERNAL tensors into the array of NamedData.
5. Pass the resolved external constants to tensor_parser, along with NDM (used for mutable external tensors).
6. Resolved external tensors are stored inside method. They are freed when the method is destructed.


Some notes:
https://docs.google.com/document/d/1_PBi4JgODuClUPD4PCUWrKNjyUH54zOUHGUJ3QHDNes/edit?tab=t.0#heading=h.blsvwraxss7g

Differential Revision: [D69477027](https://our.internmc.facebook.com/intern/diff/D69477027/)

lucylq committed:
[flat_tensor] Persist FreeableBuffers of external constants in method

## Problem
Currently, the FlatTensorDataMap persists tensors, and returns a FreeableBuffer with an empty free function.

The NamedDataMap should not persist data, as most cases (eg. delegate) will want it to be freed.

Ownership should be on the caller; `get_data` returns a FreeableBuffer that 'owns' the data. The FreeableBuffer in turn is owned by the caller.

NOTE: this doesn't support the case where we want to share plain tensors between methods/pte files at runtime. A custom NDM could support that use-case.

## This diff:
1. Introduces a 'NamedData' struct to method.h. This holds a key and a FreeeableBuffer.
2. Iterate over all the flatbuffer tensors to count the constants tagged with EXTERNAL.
NOTE: this will increase load time for all users. Potentially allocate chunks of 16 and use a linked list to store external constants, or store this number in PTE file.
3. Allocate space for num_external_constants using the method allocator.
4. Iterate over all flatbuffer tensors and use the named_data_map to resolve EXTERNAL tensors into the array of NamedData.
5. Pass the resolved external constants to tensor_parser, along with NDM (used for mutable external tensors).
6. Resolved external tensors are stored inside method. They are freed when the method is destructed.


Some notes:
https://docs.google.com/document/d/1_PBi4JgODuClUPD4PCUWrKNjyUH54zOUHGUJ3QHDNes/edit?tab=t.0#heading=h.blsvwraxss7g

Differential Revision: [D69477027](https://our.internmc.facebook.com/intern/diff/D69477027/)

[ghstack-poisoned]

lucylq committed:
Update on "[flat_tensor] Persist FreeableBuffers of external constants in method"


## Problem
Currently, the FlatTensorDataMap persists tensors, and returns a FreeableBuffer with an empty free function.

The NamedDataMap should not persist data, as most cases (eg. delegate) will want it to be freed.

Ownership should be on the caller; `get_data` returns a FreeableBuffer that 'owns' the data. The FreeableBuffer in turn is owned by the caller.

NOTE: this doesn't support the case where we want to share plain tensors between methods/pte files at runtime. A custom NDM could support that use-case.

## This diff:
1. Introduces a 'NamedData' struct to method.h. This holds a key and a FreeeableBuffer.
2. Iterate over all the flatbuffer tensors to count the constants tagged with EXTERNAL.
NOTE: this will increase load time for all users. Potentially allocate chunks of 16 and use a linked list to store external constants, or store this number in PTE file.
3. Allocate space for num_external_constants using the method allocator.
4. Iterate over all flatbuffer tensors and use the named_data_map to resolve EXTERNAL tensors into the array of NamedData.
5. Pass the resolved external constants to tensor_parser, along with NDM (used for mutable external tensors).
6. Resolved external tensors are stored inside method. They are freed when the method is destructed.


Some notes:
https://docs.google.com/document/d/1_PBi4JgODuClUPD4PCUWrKNjyUH54zOUHGUJ3QHDNes/edit?tab=t.0#heading=h.blsvwraxss7g

Differential Revision: [D69477027](https://our.internmc.facebook.com/intern/diff/D69477027/)

[ghstack-poisoned]

lucylq commented:
For: https://github.com/pytorch/executorch/issues/8393

dbort commented on runtime/executor/tensor_parser_exec_aten.cpp:
```diff
@@ -146,76 +185,43 @@ ET_NODISCARD Result<void*> getTensorDataPtr(
-          tensor_layout.dim_order()[i],
-          i);
-    }
+    const char* key =
```
Since this has a fairly long scope, `fqn` or `fully_qualified_name` would make it more self-documenting. It happens to be used as a key, but an FQN is what it _is_.

dbort commented on runtime/executor/method.cpp:
```diff
@@ -299,23 +411,33 @@ Error Method::parse_values(const NamedDataMap* named_data_map) {
 
   for (size_t i = 0; i < n_value; ++i) {
     auto serialization_value = flatbuffer_values->Get(i);
-    // Ensure that the `val_as_X()` calls will return non-null pointers.
-    ET_CHECK_OR_RETURN_ERROR(
-        serialization_value != nullptr &&
-            (serialization_value->val_type() ==
-                 executorch_flatbuffer::KernelTypes::Null ||
-             serialization_value->val() != nullptr),
-        InvalidProgram,
-        "Null value at index %" ET_PRIsize_t,
-        i);
-
```
Is this handled somewhere else now? This was a fix for crashes when loading corrupt PTEs

lucylq replied:
I added it to the first iteration through the flatbuffer values (in get_num_external_constants)

dbort reviewed (changes requested):
The overall shape of this looks good. The main issues right now:
- ~Method can look at uninitialized NamedData entries
- Some new types/functions shouldn't be part of the public API

Also, are there tests for this code path already, or are you planning to add some later?

lucylq commented on runtime/executor/tensor_parser_exec_aten.cpp:
```diff
@@ -11,6 +11,7 @@
 #include <executorch/runtime/core/evalue.h>
 #include <executorch/runtime/core/exec_aten/exec_aten.h>
 #include <executorch/runtime/executor/memory_manager.h>
+#include <executorch/runtime/executor/method.h>
```
thanks for the clarification here!

lucylq commented on runtime/executor/method.cpp:
```diff
@@ -289,6 +289,118 @@ Result<bool> parse_cond_value(const EValue& cond_value) {
+            buffer.ok(),
+            InvalidExternalData,
+            "Buffer retrieved from get_data is not valid");
+        external_constants_[index].buffer =
+            memory_manager_->method_allocator()
+                ->allocateInstance<FreeableBuffer>();
+        if (external_constants_[index].buffer == nullptr) {
+          return Error::MemoryAllocationFailed;
+        }
```
Thanks!

lucylq commented on runtime/executor/method.h:
```diff
@@ -43,6 +43,17 @@ using OpFunction = void (*)(KernelRuntimeContext&, EValue**);
 /// argument list for a single instruction
 using InstructionArgs = Span<EValue*>;
 
+/// Data structure to hold key and data buffer for external data used
+/// in a method.
+struct NamedData {
+  const char* key;
+  FreeableBuffer* buffer;
+};
```
thanks for the suggestions! Updated by forward-declaring it inside method.h and keeping the definition in tensor_parser.h

lucylq commented on runtime/executor/tensor_parser_exec_aten.cpp:
```diff
@@ -146,76 +185,43 @@ ET_NODISCARD Result<void*> getTensorDataPtr(
-          s_tensor->extra_tensor_info()->fully_qualified_name()->c_str());
-      if (!data_res.ok()) {
-        return data_res.error();
+      for (int i = 0; i < external_constants.size(); i++) {
+        if (strcmp(external_constants[i].key, key) == 0) {
+          // The const_cast is 'ok' here because program and runtime should
+          // guarantee that this data is never modified.
+          return const_cast<void*>(external_constants[i].buffer->data());
+        }
       }
```
this makes sense, thanks!

lucylq committed:
Update on "[flat_tensor] Persist FreeableBuffers of external constants in method"


## Problem
Currently, the FlatTensorDataMap persists tensors, and returns a FreeableBuffer with an empty free function.

The NamedDataMap should not persist data, as most cases (eg. delegate) will want it to be freed.

Ownership should be on the caller; `get_data` returns a FreeableBuffer that 'owns' the data. The FreeableBuffer in turn is owned by the caller.

NOTE: this doesn't support the case where we want to share plain tensors between methods/pte files at runtime. A custom NDM could support that use-case.

## This diff:
1. Introduces a 'NamedData' struct to method.h. This holds a key and a FreeeableBuffer.
2. Iterate over all the flatbuffer tensors to count the constants tagged with EXTERNAL.
NOTE: this will increase load time for all users. Potentially allocate chunks of 16 and use a linked list to store external constants, or store this number in PTE file.
3. Allocate space for num_external_constants using the method allocator.
4. Iterate over all flatbuffer tensors and use the named_data_map to resolve EXTERNAL tensors into the array of NamedData.
5. Pass the resolved external constants to tensor_parser, along with NDM (used for mutable external tensors).
6. Resolved external tensors are stored inside method. They are freed when the method is destructed.


Some notes:
https://docs.google.com/document/d/1_PBi4JgODuClUPD4PCUWrKNjyUH54zOUHGUJ3QHDNes/edit?tab=t.0#heading=h.blsvwraxss7g

Differential Revision: [D69477027](https://our.internmc.facebook.com/intern/diff/D69477027/)

[ghstack-poisoned]

lucylq commented on runtime/executor/method.cpp:
```diff
@@ -289,6 +290,109 @@ Result<bool> parse_cond_value(const EValue& cond_value) {
+  size_t n_external_constants = 0;
+  for (size_t i = 0; i < n_value; ++i) {
+    auto serialization_value = flatbuffer_values->Get(i);
+    // Ensure that the `val_as_X()` calls will return non-null pointers.
```
@dbort I moved the check here, as this is now the first iteration over the flatbuffer values.

dbort commented on runtime/executor/tensor_parser.h:
```diff
@@ -21,18 +21,34 @@ namespace executorch {
+  FreeableBuffer buffer;
+};
+
+NamedData* get_data_by_key(const char* key, Span<NamedData> external_constants);
```
This function searches a generic array of NamedData, and could be used in situations other than looking for external constants: a generic param name like `entries` would make that more clear.

dbort commented on runtime/executor/tensor_parser_exec_aten.cpp:
```diff
@@ -111,12 +111,62 @@ ET_NODISCARD Result<BoxedEvalueList<executorch::aten::Tensor>> parseTensorList(
       evalp_list, tensor_list, tensor_indices->size());
 }
 
+ET_NODISCARD Error validateExternalTensor(
```
Nothing about this is specific to external tensors per se; `validateTensorLayout` (and renaming the param to something generic like `expected_layout`) would make that more clear.

dbort commented on runtime/executor/tensor_parser_exec_aten.cpp:
```diff
@@ -111,12 +111,62 @@ ET_NODISCARD Result<BoxedEvalueList<executorch::aten::Tensor>> parseTensorList(
+ET_NODISCARD Error validateExternalTensor(
+    const executorch_flatbuffer::Tensor* s_tensor,
+    const TensorLayout& external_tensor) {
+  // Compatibility checking.
```
This is more of a comment on the entire function

dbort commented on runtime/executor/tensor_parser_exec_aten.cpp:
```diff
@@ -146,76 +196,41 @@ ET_NODISCARD Result<void*> getTensorDataPtr(
-          s_tensor->extra_tensor_info()->fully_qualified_name()->c_str());
-      if (!data_res.ok()) {
-        return data_res.error();
+      for (int i = 0; i < external_constants.size(); i++) {
+        if (strcmp(external_constants[i].key, fqn) == 0) {
+          // The const_cast is 'ok' here because program and runtime should
+          // guarantee that this data is never modified.
+          return const_cast<void*>(external_constants[i].buffer.data());
+        }
+      }
```
Could use `get_data_by_key(fqn)->buffer.data()` here (though with a check that get_data_by_key returns non-null)

lucylq replied:
thanks!

dbort reviewed (changes requested):
Thanks for the changes, looking good. The biggest remaining issue is the buffer leak when parse_values fails.

lucylq commented on runtime/executor/method.cpp:
```diff
@@ -299,23 +403,34 @@ Error Method::parse_values(const NamedDataMap* named_data_map) {
+    Result<size_t> n_resolved_constants =
+        parse_external_constants(named_data_map);
+    if (!n_resolved_constants.ok()) {
+      return n_resolved_constants.error();
```
Thanks for pointing this out! Sorry, I should have checked the value-loading loop more carefully.

lucylq committed:
Update on "[flat_tensor] Persist FreeableBuffers of external constants in method"


## Problem
Currently, the FlatTensorDataMap persists tensors, and returns a FreeableBuffer with an empty free function.

The NamedDataMap should not persist data, as most cases (eg. delegate) will want it to be freed.

Ownership should be on the caller; `get_data` returns a FreeableBuffer that 'owns' the data. The FreeableBuffer in turn is owned by the caller.

NOTE: this doesn't support the case where we want to share plain tensors between methods/pte files at runtime. A custom NDM could support that use-case.

## This diff:
1. Introduces a 'NamedData' struct to method.h. This holds a key and a FreeeableBuffer.
2. Iterate over all the flatbuffer tensors to count the constants tagged with EXTERNAL.
NOTE: this will increase load time for all users. Potentially allocate chunks of 16 and use a linked list to store external constants, or store this number in PTE file.
3. Allocate space for num_external_constants using the method allocator.
4. Iterate over all flatbuffer tensors and use the named_data_map to resolve EXTERNAL tensors into the array of NamedData.
5. Pass the resolved external constants to tensor_parser, along with NDM (used for mutable external tensors).
6. Resolved external tensors are stored inside method. They are freed when the method is destructed.


Some notes:
https://docs.google.com/document/d/1_PBi4JgODuClUPD4PCUWrKNjyUH54zOUHGUJ3QHDNes/edit?tab=t.0#heading=h.blsvwraxss7g

Differential Revision: [D69477027](https://our.internmc.facebook.com/intern/diff/D69477027/)

[ghstack-poisoned]

lucylq committed:
Update on "[flat_tensor] Persist FreeableBuffers of external constants in method"


## Problem
Currently, the FlatTensorDataMap persists tensors, and returns a FreeableBuffer with an empty free function.

The NamedDataMap should not persist data, as most cases (eg. delegate) will want it to be freed.

Ownership should be on the caller; `get_data` returns a FreeableBuffer that 'owns' the data. The FreeableBuffer in turn is owned by the caller.

NOTE: this doesn't support the case where we want to share plain tensors between methods/pte files at runtime. A custom NDM could support that use-case.

## This diff:
1. Introduces a 'NamedData' struct to method.h. This holds a key and a FreeeableBuffer.
2. Iterate over all the flatbuffer tensors to count the constants tagged with EXTERNAL.
NOTE: this will increase load time for all users. Potentially allocate chunks of 16 and use a linked list to store external constants, or store this number in PTE file.
3. Allocate space for num_external_constants using the method allocator.
4. Iterate over all flatbuffer tensors and use the named_data_map to resolve EXTERNAL tensors into the array of NamedData.
5. Pass the resolved external constants to tensor_parser, along with NDM (used for mutable external tensors).
6. Resolved external tensors are stored inside method. They are freed when the method is destructed.


Some notes:
https://docs.google.com/document/d/1_PBi4JgODuClUPD4PCUWrKNjyUH54zOUHGUJ3QHDNes/edit?tab=t.0#heading=h.blsvwraxss7g

Differential Revision: [D69477027](https://our.internmc.facebook.com/intern/diff/D69477027/)

[ghstack-poisoned]

dbort commented on runtime/executor/method.cpp:
```diff
@@ -289,6 +290,113 @@ Result<bool> parse_cond_value(const EValue& cond_value) {
+    new (&external_constants_[n_external_constants_].buffer)
+        FreeableBuffer(std::move(buffer.get()));
+
+    n_external_constants_ = i + 1;
```
Setting this to `i + 1` will cause it to skip ahead after encountering non-Tensor entries, in most cases causing an overrun of the external_constants_ array. It should only advance when adding an entry.

```suggestion
    n_external_constants_ += 1;
```

lucylq replied:
ah, oops. Thank you!

dbort commented on runtime/executor/tensor_parser.h:
```diff
@@ -21,18 +21,34 @@ namespace executorch {
     const executorch_flatbuffer::Tensor* s_tensor,
-    const NamedDataMap* named_data_map = nullptr);
+    const NamedDataMap* named_data_map = nullptr,
+    Span<NamedData> entries = {});
```
In this case `external_constants` (or whatever) was appropriate, because that's what they are. It's only get_data_by_key that is fully generic.

dbort reviewed (changes requested):
One last bug, but looks good overall. ASAN would have caught this if we had tests that exercised the new paths.

lucylq committed:
Update on "[flat_tensor] Persist FreeableBuffers of external constants in method"


## Problem
Currently, the FlatTensorDataMap persists tensors, and returns a FreeableBuffer with an empty free function.

The NamedDataMap should not persist data, as most cases (eg. delegate) will want it to be freed.

Ownership should be on the caller; `get_data` returns a FreeableBuffer that 'owns' the data. The FreeableBuffer in turn is owned by the caller.

NOTE: this doesn't support the case where we want to share plain tensors between methods/pte files at runtime. A custom NDM could support that use-case.

## This diff:
1. Introduces a 'NamedData' struct to method.h. This holds a key and a FreeeableBuffer.
2. Iterate over all the flatbuffer tensors to count the constants tagged with EXTERNAL.
NOTE: this will increase load time for all users. Potentially allocate chunks of 16 and use a linked list to store external constants, or store this number in PTE file.
3. Allocate space for num_external_constants using the method allocator.
4. Iterate over all flatbuffer tensors and use the named_data_map to resolve EXTERNAL tensors into the array of NamedData.
5. Pass the resolved external constants to tensor_parser, along with NDM (used for mutable external tensors).
6. Resolved external tensors are stored inside method. They are freed when the method is destructed.


Some notes:
https://docs.google.com/document/d/1_PBi4JgODuClUPD4PCUWrKNjyUH54zOUHGUJ3QHDNes/edit?tab=t.0#heading=h.blsvwraxss7g

Differential Revision: [D69477027](https://our.internmc.facebook.com/intern/diff/D69477027/)

[ghstack-poisoned]

lucylq committed:
Update on "[flat_tensor] Persist FreeableBuffers of external constants in method"


## Problem
Currently, the FlatTensorDataMap persists tensors, and returns a FreeableBuffer with an empty free function.

The NamedDataMap should not persist data, as most cases (eg. delegate) will want it to be freed.

Ownership should be on the caller; `get_data` returns a FreeableBuffer that 'owns' the data. The FreeableBuffer in turn is owned by the caller.

NOTE: this doesn't support the case where we want to share plain tensors between methods/pte files at runtime. A custom NDM could support that use-case.

## This diff:
1. Introduces a 'NamedData' struct to method.h. This holds a key and a FreeeableBuffer.
2. Iterate over all the flatbuffer tensors to count the constants tagged with EXTERNAL.
NOTE: this will increase load time for all users. Potentially allocate chunks of 16 and use a linked list to store external constants, or store this number in PTE file.
3. Allocate space for num_external_constants using the method allocator.
4. Iterate over all flatbuffer tensors and use the named_data_map to resolve EXTERNAL tensors into the array of NamedData.
5. Pass the resolved external constants to tensor_parser, along with NDM (used for mutable external tensors).
6. Resolved external tensors are stored inside method. They are freed when the method is destructed.


Some notes:
https://docs.google.com/document/d/1_PBi4JgODuClUPD4PCUWrKNjyUH54zOUHGUJ3QHDNes/edit?tab=t.0#heading=h.blsvwraxss7g

Differential Revision: [D69477027](https://our.internmc.facebook.com/intern/diff/D69477027/)

[ghstack-poisoned]
